File: mobile/ios/ActiveChainWalletApp/Sources/NetworkSelection.swift
import Foundation
import Combine

public struct NetworkProfile: Identifiable, Equatable {
    public let id: String
    public let displayName: String
    public let genesis: String
    public let rpcURL: URL
    public let faucetURL: URL?
    public let assets: [String]
}

public final class NetworkSelection: ObservableObject {
    @Published public private(set) var selected: NetworkProfile
    @Published public private(set) var visibleAssets: [String]
    private let profiles: [NetworkProfile]

    public init(profiles: [NetworkProfile], selectedID: String? = nil) {
        precondition(!profiles.isEmpty)
        self.profiles = profiles
        let initial = profiles.first { $0.id == selectedID } ?? profiles[0]
        self.selected = initial
        self.visibleAssets = initial.assets
    }

    public func switchTo(_ id: String) -> Bool {
        guard let next = profiles.first(where: { $0.id == id }) else { return false }
        selected = next
        visibleAssets = next.assets
        return true
    }
}

File: mobile/android/app/src/main/kotlin/dev/activechain/wallet/NetworkSelection.kt
package dev.activechain.wallet

data class NetworkProfile(val id: String, val displayName: String, val genesis: String,
                          val rpcUrl: String, val faucetUrl: String?, val assets: List<String>)

class NetworkSelection(profiles: List<NetworkProfile>, selectedId: String? = null, private val preferences: android.content.SharedPreferences? = null) {
    private val all = profiles.associateBy { it.id }
    var selected: NetworkProfile = all[selectedId ?: preferences?.getString("activechain.selected-network", null)] ?: profiles.first()
        private set
    val visibleAssets: List<String> get() = selected.assets
    fun switchTo(id: String): Boolean { val next = all[id] ?: return false; selected = next; preferences?.edit()?.putString("activechain.selected-network", id)?.apply(); return true }
}
